File: clients/go/webhook_test.go
package crucible_test

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"strings"
	"testing"
	"time"

	crucible "github.com/Unluckyathecking/crucible/clients/go"
)

// testSign replicates gateway/internal/webhookout.Sign locally so tests build
// the positive vector without importing the gateway package tree.
func testSign(secret []byte, timestamp string, body []byte) string {
	mac := hmac.New(sha256.New, secret)
	mac.Write([]byte(timestamp))
	mac.Write([]byte("."))
	mac.Write(body)
	return hex.EncodeToString(mac.Sum(nil))
}

func TestVerifyWebhook_valid(t *testing.T) {
	secret := make([]byte, 32)
	for i := range secret {
		secret[i] = byte(i + 1)
	}
	secretHex := hex.EncodeToString(secret)
	body := []byte(`{"event":"delivery.succeeded","data":{"id":1}}`)
	ts := fmt.Sprintf("%d", time.Now().Unix())
	sig := testSign(secret, ts, body)
	header := "t=" + ts + ",v1=" + sig

	if err := crucible.VerifyWebhook(secretHex, header, body, 5*time.Minute); err != nil {
		t.Fatalf("VerifyWebhook: %v", err)
	}
}

func TestVerifyWebhook_tamperedBody(t *testing.T) {
	secret := make([]byte, 32)
	secretHex := hex.EncodeToString(secret)
	body := []byte(`{"event":"original"}`)
	ts := fmt.Sprintf("%d", time.Now().Unix())
	sig := testSign(secret, ts, body)
	header := "t=" + ts + ",v1=" + sig

	err := crucible.VerifyWebhook(secretHex, header, []byte(`{"event":"tampered"}`), 5*time.Minute)
	if err == nil {
		t.Fatal("expected error for tampered body, got nil")
	}
}

func TestVerifyWebhook_wrongSecret(t *testing.T) {
	correctSecret := make([]byte, 32)
	for i := range correctSecret {
		correctSecret[i] = 0xAA
	}
	wrongSecret := make([]byte, 32)
	for i := range wrongSecret {
		wrongSecret[i] = 0xBB
	}
	secretHex := hex.EncodeToString(correctSecret)
	body := []byte(`{"event":"test"}`)
	ts := fmt.Sprintf("%d", time.Now().Unix())
	sig := testSign(wrongSecret, ts, body)
	header := "t=" + ts + ",v1=" + sig

	if err := crucible.VerifyWebhook(secretHex, header, body, 5*time.Minute); err == nil {
		t.Fatal("expected error for wrong secret, got nil")
	}
}

func TestVerifyWebhook_expiredTimestamp(t *testing.T) {
	secret := make([]byte, 32)
	secretHex := hex.EncodeToString(secret)
	body := []byte(`{"event":"test"}`)
	old := time.Now().Add(-10 * time.Minute)
	ts := fmt.Sprintf("%d", old.Unix())
	sig := testSign(secret, ts, body)
	header := "t=" + ts + ",v1=" + sig

	if err := crucible.VerifyWebhook(secretHex, header, body, 5*time.Minute); err == nil {
		t.Fatal("expected error for expired timestamp, got nil")
	}
}

func TestVerifyWebhook_multipleV1Candidates_secondValid(t *testing.T) {
	secret := make([]byte, 32)
	for i := range secret {
		secret[i] = 0x42
	}
	secretHex := hex.EncodeToString(secret)
	body := []byte(`{"event":"multi"}`)
	ts := fmt.Sprintf("%d", time.Now().Unix())
	validSig := testSign(secret, ts, body)
	invalidSig := strings.Repeat("a", 64)
	header := "t=" + ts + ",v1=" + invalidSig + ",v1=" + validSig

	if err := crucible.VerifyWebhook(secretHex, header, body, 5*time.Minute); err != nil {
		t.Fatalf("VerifyWebhook with multiple candidates: %v", err)
	}
}

func TestVerifyWebhook_boundedCandidates(t *testing.T) {
	secret := make([]byte, 32)
	secretHex := hex.EncodeToString(secret)
	body := []byte(`{"event":"test"}`)
	ts := fmt.Sprintf("%d", time.Now().Unix())
	validSig := testSign(secret, ts, body)

	parts := []string{"t=" + ts}
	for i := 0; i < 8; i++ {
		parts = append(parts, "v1="+strings.Repeat("b", 64))
	}
	parts = append(parts, "v1="+validSig)
	header := strings.Join(parts, ",")

	if err := crucible.VerifyWebhook(secretHex, header, body, 5*time.Minute); err == nil {
		t.Fatal("expected error when valid sig is beyond maxSigCandidates, got nil")
	}
}

func TestVerifyWebhook_missingHeader(t *testing.T) {
	if err := crucible.VerifyWebhook("aabb", "", []byte("body"), 5*time.Minute); err == nil {
		t.Fatal("expected error for missing header, got nil")
	}
}

File: clients/typescript/test/webhook.test.ts
import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { createHmac } from "node:crypto";
import {
  verifyWebhook,
  WebhookVerificationError,
  DEFAULT_TOLERANCE_MS,
} from "../src/webhook";

// testSign replicates gateway/internal/webhookout.Sign so tests build the
// positive vector without importing the gateway package tree.
function testSign(secret: Buffer, timestamp: string, body: Buffer): string {
  const mac = createHmac("sha256", secret);
  mac.update(timestamp);
  mac.update(".");
  mac.update(body);
  return mac.digest("hex");
}

const secret = Buffer.alloc(32, 0x42);
const secretHex = secret.toString("hex");
const body = Buffer.from('{"event":"delivery.succeeded","data":{"id":1}}');
const ts = Math.floor(Date.now() / 1000).toString();

describe("verifyWebhook", () => {
  it("verifies a valid signature (positive vector, matches gateway Sign)", () => {
    const sig = testSign(secret, ts, body);
    const header = `t=${ts},v1=${sig}`;
    assert.doesNotThrow(() => verifyWebhook(secretHex, header, body));
  });

  it("accepts body as string (auto-encodes to Buffer before hashing)", () => {
    const bodyStr = '{"event":"string-body"}';
    const bodyBuf = Buffer.from(bodyStr);
    const sig = testSign(secret, ts, bodyBuf);
    const header = `t=${ts},v1=${sig}`;
    assert.doesNotThrow(() => verifyWebhook(secretHex, header, bodyStr));
  });

  it("rejects tampered body", () => {
    const sig = testSign(secret, ts, body);
    const header = `t=${ts},v1=${sig}`;
    assert.throws(
      () => verifyWebhook(secretHex, header, Buffer.from('{"event":"tampered"}')),
      (err: unknown) => {
        assert.ok(err instanceof WebhookVerificationError, `expected WebhookVerificationError, got ${err}`);
        return true;
      },
    );
  });

  it("rejects wrong secret", () => {
    const wrongSecret = Buffer.alloc(32, 0xff);
    const sig = testSign(wrongSecret, ts, body);
    const header = `t=${ts},v1=${sig}`;
    assert.throws(
      () => verifyWebhook(secretHex, header, body),
      (err: unknown) => {
        assert.ok(err instanceof WebhookVerificationError);
        return true;
      },
    );
  });

  it("rejects expired timestamp", () => {
    const oldTs = Math.floor((Date.now() - 10 * 60 * 1000) / 1000).toString();
    const sig = testSign(secret, oldTs, body);
    const header = `t=${oldTs},v1=${sig}`;
    assert.throws(
      () => verifyWebhook(secretHex, header, body, 5 * 60 * 1000),
      (err: unknown) => {
        assert.ok(err instanceof WebhookVerificationError);
        return true;
      },
    );
  });

  it("accepts second v1= candidate when first is invalid", () => {
    const validSig = testSign(secret, ts, body);
    const invalidSig = "a".repeat(64);
    const header = `t=${ts},v1=${invalidSig},v1=${validSig}`;
    assert.doesNotThrow(() => verifyWebhook(secretHex, header, body));
  });

  it("rejects valid sig placed past maxSigCandidates bound (9th of 8)", () => {
    const validSig = testSign(secret, ts, body);
    const fakeSigs = Array<string>(8)
      .fill("b".repeat(64))
      .map((s) => `v1=${s}`)
      .join(",");
    const header = `t=${ts},${fakeSigs},v1=${validSig}`;
    assert.throws(
      () => verifyWebhook(secretHex, header, body),
      (err: unknown) => {
        assert.ok(err instanceof WebhookVerificationError);
        return true;
      },
    );
  });

  it("throws on missing header", () => {
    assert.throws(
      () => verifyWebhook(secretHex, "", body),
      (err: unknown) => {
        assert.ok(err instanceof WebhookVerificationError);
        return true;
      },
    );
  });

  it("exposes DEFAULT_TOLERANCE_MS as 5 minutes", () => {
    assert.equal(DEFAULT_TOLERANCE_MS, 5 * 60 * 1000);
  });
});
